wtxmgr/db.go

```diff
@@ -0,0 +1,1402 @@
+	}
+	return nil
+}
+
+func existsRawUnminedInput(ns walletdb.Bucket, k []byte) (v []byte) {
+	return ns.Bucket(bucketUnminedInputs).Get(k)
+}
+
+func deleteRawUnminedInput(ns walletdb.Bucket, k []byte) error {
+	err := ns.Bucket(bucketUnminedInputs).Delete(k)
+	if err != nil {
+		str := "failed to delete unmined input"
+		return storeError(ErrDatabase, str, err)
+	}
+	return nil
+}
+
+// openStore opens an existing transaction store from the passed namespace.  If
+// necessary, an already existing store is upgraded to newer db format.
+func openStore(namespace walletdb.Namespace) error {
+	var version uint32
+	err := scopedView(namespace, func(ns walletdb.Bucket) error {
+		// Verify a store already exists and upgrade as necessary.
+		v := ns.Get(rootVersion)
+		if len(v) != 4 {
+			return nil
+		}
+		version = byteOrder.Uint32(v)
+		return nil
+	})
+	if err != nil {
+		const desc = "failed to open existing store"
+		if serr, ok := err.(Error); ok {
+			serr.Desc = desc + ": " + serr.Desc
+			return serr
+		}
+		return storeError(ErrDatabase, desc, err)
+	}
+
+	// The initial version is one.  If no store exists and no version was
+	// saved, this variable will be zero.
+	if version == 0 {
+		str := "no transaction store exists in namespace"
+		return storeError(ErrNoExists, str, nil)
+	}
+
+	// Cannot continue if the saved database is too new for this software.
+	// This probably indicates an outdated binary.
+	if version > LatestVersion {
+		str := fmt.Sprintf("recorded version %d is newer that latest "+
+			"understood version %d", version, LatestVersion)
+		return storeError(ErrUnknownVersion, str, nil)
+	}
+
+	// Upgrade the tx store as needed, one version at a time, until
+	// LatestVersion is reached.  Versions are not skipped when performing
+	// database upgrades, and each upgrade is done in its own transaction.
+	//
+	// No upgrades yet.
+	//if version < LatestVersion {
+	//	err := scopedUpdate(namespace, func(ns walletdb.Bucket) error {
+	//	})
+	//	if err != nil {
+	//		// Handle err
+	//	}
+	//}
+
+	return nil
+}
+
+// createStore creates the tx store (with the latest db version) in the passed
+// namespace.  If a store already exists, ErrAlreadyExists is returned.
+func createStore(namespace walletdb.Namespace) error {
+	// Initialize the buckets and root bucket fields as needed.
+	err := scopedUpdate(namespace, func(ns walletdb.Bucket) error {
+		// Ensure that nothing currently exists in the namespace bucket.
+		ck, cv := ns.Cursor().First()
+		if ck != nil || cv != nil {
+			const str = "namespace is not empty"
+			return storeError(ErrAlreadyExists, str, nil)
+		}
+
+		// Write the latest store version.
+		v := make([]byte, 4)
+		byteOrder.PutUint32(v, LatestVersion)
+		err := ns.Put(rootVersion, v)
+		if err != nil {
+			str := "failed to store latest database version"
+			return storeError(ErrDatabase, str, err)
+		}
+
+		// Save the creation date of the store if it does not already
+		// exist.
+		v = ns.Get(rootCreateDate)
```

Comment:
No need to check for this.  The code above already proved the namespace is empty.